### DISCUSSION TITLE
Problems transforming tractogram

### DISCUSSION THREAD

**Discussion by karianneksw** [dipy/dipy#3354] [Registration]
We are having difficulty applying a transform to a tck file and saving it out in vtk format in DIPY. The transform was verified using Slicer but when applying this transform in DIPY and saving it out as a vtk, the output is inaccurate (slight left-right shift) and also loads in Slicer in LPS rather than RAS space. Below is our example code. Do you know what we are doing incorrectly?

```
import SimpleITK as sitk

tractogram_in = load_tractogram(tck_in_file, ref_img_acpc)
streamlines_in = tractogram_in.streamlines

xfm_in = sitk.ReadTransform(xfm_DTI_to_acpc)
xfm_in = sitk.AffineTransform(xfm_in)
xfm_m = np.eye(4)
xfm_m[0:3,0:3] = np.asarray(xfm_in.GetMatrix()).reshape((3,3))
xfm_m[0:3,3] = xfm_in.GetTranslation()

xfm_m_inv = np.linalg.inv(xfm_m)
streamlines_out = transform_streamlines(streamlines_in,xfm_m_inv)

tractogram_out = StatefulTractogram(streamlines_out, ref_img_dti, Space("rasmm"))

save_tractogram(tractogram_out,tck_out_file)```

**Comment by skoudoro** [comment]
Hi @karianneksw

Apologize in advance for basic questions but your problem can come from multiple places and it will be very hard to help without many details.

> tractogram_in = load_tractogram(tck_in_file, ref_img_acpc)

- in which space is `tck_in_file`? how did you generate and save it ? is it rasmm?  in which space is `ref_img_acpc`. 
- This already can change your data, a space transformation will be done. that's why we recommend `*.trx` file format or `*.trk` file format since they save the space of your streamlines, no surprise.

> xfm_in = sitk.ReadTransform(xfm_DTI_to_acpc)

`xfm_DTI_to_acpc` has been created with sitk I suppose.  What is it exactly ? does it come from 2 images registered or streamlines registered? this will have an impact on `np.linalg.inv(xfm_m)`.  (necessary if it is transformation from 2 images/volumes registered, Unecessary if it is the result of 2 tractograms registered) 

> StatefulTractogram(streamlines_out, ref_img_dti, Space("rasmm"))

 in which space is ref_img_dti ? Why do you need to precise `Space("rasmm")`

Also,
- `dipy_info my_files.trk my_file.nii.gz`  will tell you in space you are. Please, provide these details, it can help us.
- I recommend you to look at this tutorial https://docs.dipy.org/stable/examples_built/registration/streamline_registration.html

I hope it helps.

**Comment by karianneksw** [comment]
Hi @skoudoro 

Thanks for your response. I have included some answers to your questions.To provide some context, we are trying to transform a tck file from acpc to DTI space and save it out as a vtk.

- in which space is tck_in_file? how did you generate and save it ? is it rasmm? in which space is ref_img_acpc.
tck_in_file is generated using mrtrix and is saved in ACPC space. ref_img_acpc is in ACPC space as well.

- in which space is ref_img_dti ? Why do you need to precise Space("rasmm")
ref_img_dti is in dti space

 I'll have our colleague respond to the remaining unanswered questions.

**Comment by hbraunDSP** [comment]
Hi @skoudoro, I'm a co-worker of @karianneksw and I'll help to answer your questions as well as I can. Basically, we have a known-good affine transform which behaves correctly in Slicer. When we attempt to apply the same transform to a .tck file using DIPY, the output is off by about 1-10 mm, and also flipped A-P and R-L when loaded in Slicer.

> in which space is tck_in_file?

 - `tck_in_file` is generated with MRTrix and it is in RASMM space. When loaded with Slicer (using the SlicerDMRI extension) it is aligned with ref_img_acpc. MRTrix does not support .trk files so we are stuck with .tck.

> xfm_DTI_to_acpc has been created with sitk I suppose. What is it exactly ?
- The affine transform `xfm_DTI_to_acpc` is saved in a text format used by many different ITK-based programs, including Slicer and ANTS. We are using the SimpleITK package to load the transform and extract its matrix representation. This is fairly routine for us and we have used it successfully before.
- The transform was created by registering image volumes. We are aware that a transform created by registering images must be inverted when applying to a point set such as streamlines. We are confident we're not accidentally applying the inverse of the transform we want. The transform operates correctly in Slicer.

- ref_img_dti is an unregistered B0 diffusion volume. We ran tractography on ACPC-aligned images and now we have a need to transform the tract files to match the un-aligned diffusion images. 

- I suppose we wouldn't actually have to specify `Space('rasmm')` since that's the default option. It doesn't hurt though, does it?

I'll add some questions of my own:

- `tractogram_in` and `tractogram_out` both show an "LAS" voxel order if we run `print()` on them. Does that mean that the streamline data are in LAS rather than RAS space?
- if we simply load a .tck file and then save the same streamlines in .vtk format, we encounter an LPS > RAS discrepancy when loading into Slicer (i.e. R-L and A-P axes are flipped). What orientation conventions does DIPY follow when saving `.vtk` files?

**Reply by skoudoro** [reply]
> I suppose we wouldn't actually have to specify Space('rasmm') since that's the default option. It doesn't hurt though, does it?

No,

> tractogram_in and tractogram_out both show an "LAS" voxel order if we run print() on them. Does that mean that the streamline data are in LAS rather than RAS space?

Yes, and it is very strange since you said that `tck_in_file` and `ref_img_acpc` are in RAS. This is the first thing I will check.  if they are both in LAS, it make sense that they are aligned correctly in slicer.
you can check with the command line `dipy_info` or with any other software. 

> if we simply load a .tck file and then save the same streamlines in .vtk format, we encounter an LPS > RAS discrepancy when loading into Slicer (i.e. R-L and A-P axes are flipped). What orientation conventions does DIPY follow when saving .vtk files?

Since VTK files does not save the space, we do not force any space. if the streamlines/tractogram was in LAS space, the vtk file will be in LAS space. Same if it was in RAS.


You code above seems correct after all the explanation. Your print concerning LAS/RAS might be your initial issue and needs to be clarify. I would double check `ref_img_acpc` space `ref_img_dti` space  and the data used to generate `tck_in_file` space.
